gp/cartesian_graph.py

```diff
@@ -216,17 +222,21 @@ def _format_output_str_of_all_nodes(self):
             for node in active_nodes[hidden_column_idx]:
                 node.format_output_str(self)
 
-    def _fill_parameter_values(self, func_str, parameter_names_to_values):
+    def _fill_parameter_values(
+        self, func_str: str, parameter_names_to_values: Optional[Dict[str, float]] = None
+    ) -> str:
         g = re.findall("<p[0-9]+>", func_str)
-        if g and parameter_names_to_values is None:
+        if len(g) != 0 and parameter_names_to_values is None:
             raise ValueError("parameter node found but no value dict provided")
-        for parameter_name in g:
-            func_str = func_str.replace(
-                parameter_name, str(parameter_names_to_values[parameter_name])
-            )
+        # explicit elif branch is necessary because mypy does not pass otherwise
```

Comment:
this is a very strange construction. what's wrong with this according to mypy?

Reply:
`mypy` doesn't recognize the fact that we never the for-loop if `len(g) == 0` and then complains that the for-loop will break if `parameter_names_to_values = None`.